### ISSUE TITLE
Couldn't get ALEFix to work with autopep8 and black

### ISSUE THREAD

**Issue by baskarangappan** [dense-analysis/ale#3395]
**VIM version**
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Aug 22 2017 04:20:17)
Included patches: 1-670

Base platform OS: FreeBSD 10.4

## What went wrong
I've installed Ale and installed pylint and got the lining work properly. But couldn't get ALEFix to work with autopep8. Then tried installing 'black' and didn't work either. It doesn't throw any error as well.

## Reproducing the bug
Written a faulty python file, with an unused import, and tried fixing it using ALEFix and didn't work. Same goes for lengthy line or any other python related error.

in my ~/.vimrc
I added the following lines and it didn't work
" Check Python files with flake8 and pylint.
let g:ale_linters = {'python': ['flake8', 'pylint']}
" Fix Python files with black and autopep8.
let g:ale_fixers = {'python': ['black', 'autopep8']}

Then tried a suggestion which was given on the other thread, by creating a python.vim under ~/.vim/ftplugin/python.vim, and pasted below lines and (commented the above global one) and it didn't still work
" Check Python files with flake8 and pylint.
let b:ale_linters = {'python': ['flake8', 'pylint']}
" Fix Python files with black and autopep8.
let b:ale_fixers = {'python': ['black', 'autopep8']}

### :ALEInfo

 Current Filetype: python
Available Linters: ['bandit', 'flake8', 'mypy', 'prospector', 'pycodestyle', 'pydocstyle', 'pyflakes', 'pylama', 'pylint', 'pyls', 'pyre', 'pyright', 'vulture']
  Enabled Linters: ['flake8', 'pylint']
  Ignored Linters: []
 Suggested Fixers:
  'add_blank_lines_for_python_control_statements' - Add blank lines before control statements.
  'autopep8' - Fix PEP8 issues with autopep8.
  'black' - Fix PEP8 issues with black.
  'isort' - Sort Python imports with isort.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'reorder-python-imports' - Sort Python imports with reorder-python-imports.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'yapf' - Fix Python files with yapf.
 Linter Variables:
let g:ale_python_auto_pipenv = 0
let g:ale_python_flake8_auto_pipenv = 0
let g:ale_python_flake8_change_directory = 'project'
let g:ale_python_flake8_executable = 'flake8'
let g:ale_python_flake8_options = ''
let g:ale_python_flake8_use_global = 0
let g:ale_python_pylint_auto_pipenv = 0
let g:ale_python_pylint_change_directory = 1
let g:ale_python_pylint_executable = 'pylint'
let g:ale_python_pylint_options = ''
let g:ale_python_pylint_use_global = 0
let g:ale_python_pylint_use_msg_id = 0
 Global Variables:
let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_disable_lsp = 0
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'python': ['black', 'autopep8']}
let b:ale_fixers = {'python': ['black', 'autopep8']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}
let g:ale_linters = {'python': ['flake8', 'pylint']}
let b:ale_linters = {'python': ['flake8', 'pylint']}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:
(finished - exit code 0) ['/usr/local/bin/bash', '-c', '''pylint'' --version']
<<<OUTPUT STARTS>>>
pylint 2.6.0
astroid 2.4.2
Python 3.6.1 (default, Aug 12 2017, 01:13:43)
[GCC 4.2.1 Compatible FreeBSD Clang 3.4.1 (tags/RELEASE_34/dot1-final 208032)]
<<<OUTPUT ENDS>>>
(finished - exit code 0) ['/usr/local/bin/bash', '-c', '''pylint'' --version']
<<<OUTPUT STARTS>>>
pylint 2.6.0
astroid 2.4.2
Python 3.6.1 (default, Aug 12 2017, 01:13:43)
[GCC 4.2.1 Compatible FreeBSD Clang 3.4.1 (tags/RELEASE_34/dot1-final 208032)]
<<<OUTPUT ENDS>>>
(executable check - success) flake8
(started) ['/usr/local/bin/bash', '-c', '''flake8'' --version']
(executable check - success) pylint
(started) ['/usr/local/bin/bash', '-c', 'cd ''/data/home/bangappa/work/ff/ap/feature_flags'' && ''pylint'' --output-format text --msg-template="{path}:{line}:{column}: {msg_id} ({symbol}) {msg}" --reports n --from-stdin ''/data/home/bangappa/work/ff/ap/feature_fla
gs/feature_flags/feature_flags_config_manager.py'' < ''/tmp/vPQmEOK/1/feature_flags_config_manager.py''']
(finished - exit code 0) ['/usr/local/bin/bash', '-c', '''flake8'' --version']
<<<OUTPUT STARTS>>>
3.8.4 (mccabe: 0.6.1, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 2.7.13 on FreeBSD
<<<OUTPUT ENDS>>>

**Comment by aadibajpai** [comment]
I'm wondering where the problem lies—does `:ALEFix black` work for you?

**Comment by baskarangappan** [comment]
No, it doesn't work. Couldn't get ALEFix work still

**Comment by gennaro-tedesco** [comment]
I confirm the issue as well and have tried to reproduce running `:ALEFix black` too. It does nothing, same goes if using `yapf`, so the problem is clearly not with the fixer in use.

**Comment by hsanson** [comment]
Tested with this sample hello.py file:

```python
"""
Hello sample
"""
def say(msg):
    print("Hello %s" % msg)

say("World!")
```

The flake8 linter tells me that there must be two empty lines before and after the function definition.  Then running both `ALEFix black` or `ALEFix yafp` properly fixes the linter errors. The resulting buffer is:

```python
"""
Hello sample
"""


def say(msg):
    print("Hello %s" % msg)


say("World!")
```

From this test it seems, to me, that the fixers are working properly. Note I have very limited experience with Python code and am unaware of how it should look properly formatted. I do see that the fixers do modify the buffer and that linter errors are cleared afterwards.

**Comment by sonac** [comment]
You might not have your black in your path and therefore ALE can't catch it up, try running `which black`, dunno if there is the possibility to configure ALE to run `python -m black` instead of `black` though

**Comment by silentjay** [comment]
@sonac suggestion worked for me. I don't know what happened as Black was working previously, somehow must have got uninstalled. Just run `pip install black` and job done.

**Comment by hsanson** [comment]
ALE seems to work properly so will close this issue. Re-open if the issue persists.
